src/dag/write.rs

```diff
@@ -1,55 +1,195 @@
-use super::chunk::Chunk;
 use super::key::Key;
+use super::{chunk::Chunk, meta_generated::meta};
 use super::{read, Result};
 use crate::kv;
+use async_recursion::async_recursion;
+use async_std::sync::RwLock;
+use futures::future::try_join_all;
+use futures::try_join;
+use std::collections::HashSet;
+
+#[derive(Debug, Default)]
+struct HeadChange(Option<String>, Option<String>);
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+enum RefCount {
+    Inc = 1,
+    Dec = -1,
+}
 
 pub struct Write<'a> {
     kvw: Box<dyn kv::Write + 'a>,
+    changed_heads: RwLock<Vec<HeadChange>>,
+    mutated_chunks: RwLock<HashSet<String>>,
 }
 
 impl<'a> Write<'_> {
     pub fn new(kvw: Box<dyn kv::Write + 'a>) -> Write {
-        Write { kvw }
+        Write {
+            kvw,
+            changed_heads: Default::default(),
```

Comment:
[taste]: I think more clear to just construct the write value, unless there is some reason for the genericism of `Default` necessary.

Reply:
less code duplication